### PR TITLE
gh-88013: ntpath.realpath may raise TypeError in some cases

### PR DESCRIPTION
**Pull request by aisk**
* Issue: gh-88013

### CODE REVIEW

File: Lib/test/test_ntpath.py [hunk 1]
```diff
@@ -374,6 +374,10 @@ def test_realpath_basic(self):
         self.assertPathEqual(ntpath.realpath(os.fsencode(ABSTFN + "1")),
                          os.fsencode(ABSTFN))
 
+        # gh-88013: call ntpath.realpath with binary drive name may raise a
+        # TypeError. The drive should not exist to reproduce the bug.
+        self.assertEqual(ntpath.realpath(b"n:/"), b"n:\\")
```
Comment (barneygale) [inline Lib/test/test_ntpath.py]: You should be able to use [`os.listdrives()`](https://docs.python.org/3.12/library/os.html#os.listdrives) to find an unused drive.

File: Misc/NEWS.d/next/Windows/2023-03-18-21-38-00.gh-issue-88013.Z3loxC.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fixed a bug which will cause ``TypeError`` when calling
+``ntpath.realpath()`` with bytes parameter in some cases.
```
Comment (barneygale) [inline Misc/NEWS.d/next/Windows/2023-03-18-21-38-00.gh-issue-88013.Z3loxC.rst]:
```suggestion
Fixed a bug where :exc:`TypeError` was raised when calling
:func:`ntpath.realpath` with a bytes parameter in some cases.
```

File: Lib/test/test_ntpath.py (hunk 1 above)
Comment (aisk) [reply Lib/test/test_ntpath.py]: Updated, thanks for sugsestion!

**Review by zooba — COMMENTED** [review]
LGTM. Barney can merge now ;)

By the way, `listdrives` was only added in 3.12, so if we want to backport this we'll need an updated test.

File: Lib/test/test_ntpath.py [hunk 3]
```diff
@@ -374,6 +374,12 @@ def test_realpath_basic(self):
         self.assertPathEqual(ntpath.realpath(os.fsencode(ABSTFN + "1")),
                          os.fsencode(ABSTFN))
 
+        # gh-88013: call ntpath.realpath with binary drive name may raise a
+        # TypeError. The drive should not exist to reproduce the bug.
+        drives = {f"{chr(x)}:\\" for x in range(65, 91)} - set(os.listdrives())
```
Comment (barneygale) [inline Lib/test/test_ntpath.py]:
```suggestion
        drives = {f"{c}:\\" for c in string.ascii_uppercase} - set(os.listdrives())
```

(will also need an `import string` at the top)

**Comment by zooba** [comment]
@aisk No need to force push - we will always squash merge and rewrite the commit message here, and force pushes mean we can't see what was changed since the last review.

**Comment by aisk** [comment]
I just force push between reviews for clean

> @aisk No need to force push - we will always squash merge and rewrite the commit message here, and force pushes mean we can't see what was changed since the last review.

Got it

**Review by barneygale — APPROVED** [review]
LGTM!

**Comment by hugovk** [comment]
I removed the 3.10 backport label because it's now getting security fixes only.

@barneygale Would you like to do the honours with the big green merge button? :)

**Pull request merged**